### PR TITLE
[CP] Rewrite ring attention backward algorithm and enablement APIs

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132820
* #132442
* __->__ #131351

**What does this PR achieve**
1. This PR rewrite ring attention backward algorithm to fuse the alltoall and overlap the gradient communication with computation.

2. Enables memory efficient attention with CP by templating the ring attention backward to verify the accuracy as fp32 gives us higher confident about the implementation correctness.

3. Provides some experimental APIs to enable context parallelism.

4. Ensures CP work with torch.compiler. The combination of causal masking and torch.compiler has not
yet worked.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: test/distributed/_tensor/test_attention.py [hunk 1]
```diff
@@ -40,137 +38,146 @@
...
         nheads = 8
-        query = torch.rand(
+        torch.manual_seed(10)
+        dtype = (
+            torch.bfloat16 if backend == SDPBackend.FLASH_ATTENTION else torch.float32
+        )
+
+        if is_causal and compiled and self.world_size > 2:
+            # TODO: Fix this after we move `wait_tensor` to use `with_effect`.
+            return
```
Comment (wanchaol) [inline test/distributed/_tensor/test_attention.py:80]: do we know what kind of error it throwing? if so we could use `self.assertRaises` instead

File: test/distributed/_tensor/test_attention.py [hunk 2]
```diff
@@ -40,137 +38,146 @@
...
-        dkey = distribute_tensor(key, device_mesh, context_placement)
-        dvalue = distribute_tensor(value, device_mesh, context_placement)
-        for t in [dkey, dvalue]:
-            self.assertEqual(
-                t.shape, (bs, nheads, context_tokens * self.world_size, dim)
-            )
-            self.assertEqual(t.to_local().shape, (bs, nheads, context_tokens, dim))
+        # Ensure all ranks have the same initialization data.
+        with torch.no_grad():
+            dist.broadcast(q, src=0)
```
Comment (wanchaol) [inline test/distributed/_tensor/test_attention.py:103]: iiuc, here we are switching to testing the ring attention without sharding the inputs, to verify the local sdpa is numerically the same with local ring attention? 🤔 but the ring attention works in the way that it's calculating the global attention score that shards qkv on sequence dimension already, so is this comparison something we want to do?

I thought the old way of creating global qkv, shard them on sequence_dim, then comparing global spda results with ring attention should be the correct numeric test setup?

File: torch/distributed/distributed_c10d.py [hunk 3]
```diff
@@ -4943,7 +4943,7 @@ def _find_or_create_pg_by_ranks_and_tag(
                 my_ranks = rank_set
         assert my_ranks is not None, "rankset doesn't include the current node"
 
-    my_ranks.sort()
```
Comment (wanchaol) [inline torch/distributed/distributed_c10d.py:4946]: any difference between these two calls?

File: torch/distributed/_tensor/experimental/attention.py [hunk 4]
```diff
@@ -634,27 +662,198 @@ def backward_hook(grad: torch.Tensor) -> None:
...
+    buffer: torch.Tensor, mesh: DeviceMesh, seq_dim: int
+) -> torch.Tensor:
+    return buffer.chunk(mesh.size(), dim=seq_dim)[mesh.get_local_rank()]
+
+
+@contextlib.contextmanager
+@torch.no_grad()
+def context_parallel_buffers(
+    mesh: Optional[DeviceMesh],
+    buffers: List[torch.Tensor],
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/attention.py:824]: Could you add documentation on this API, specifically I am confused about what are the meanings of `buffers`, and `restore_funcs`?

File: torch/distributed/_tensor/experimental/attention.py [hunk 5]
```diff
@@ -361,153 +332,210 @@ def sdpa_backward_handler(
...
 
 customized_ops = {
     aten._scaled_dot_product_flash_attention.default: sdpa_handler,
     aten._scaled_dot_product_flash_attention_backward.default: sdpa_backward_handler,
+    aten._scaled_dot_product_efficient_attention.default: sdpa_handler,
+    aten._scaled_dot_product_efficient_attention_backward.default: sdpa_backward_handler,
 }
 
 
+# Following experimental APIs allow users to enable CP.
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/attention.py:536]: looks like we have multiple user facing APIs: 1. attention_context_parallel 2. enable_context_parallel 3. context_parallel_buffers. It would be good to consolidate and remove some that are not necessary

File: torch/distributed/_tensor/experimental/attention.py [hunk 6]
```diff
@@ -361,153 +332,210 @@ def sdpa_backward_handler(
...
-
     next_kv = None
+    next_grad_kv = None
+    rest: List[Any]
+    grad_query_, grad_key_, grad_value_ = None, None, None
 
-    out_grad_queries = []
-    out_grad_keys = []
-    out_grad_values = []
+    grad_query = torch.zeros_like(query, dtype=torch.float32)
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/attention.py:375]: curious are the fp32 grad_query/key/value intentional?

**Review by wanchaol — COMMENTED** [review]
Have some comments and suggestions inlined

File: test/distributed/_tensor/test_attention.py (hunk 1 above)
Comment (fegin) [reply test/distributed/_tensor/test_attention.py:80]: The error can be random, either accuracy or timeout. So we could not use `assertRaises` at this moment.

File: test/distributed/_tensor/test_attention.py (hunk 2 above)
Comment (fegin) [reply test/distributed/_tensor/test_attention.py:103]: There should be no differences between these two tests. The way current test is to get the global SDPA result and rank0 gets the first half and rank1 gets the second half. This should match the local SDPA result using because we shard on the sequence dimension.

But ye, it may be more straightforward to just compare the global results but that requires the reshuffling which is introduced in the next PR. The test will be changed to compare the global SDPA result in the next PR. I don't want to merge everything into one PR, so bear with this test for just a while.

File: torch/distributed/_tensor/experimental/attention.py (hunk 6 above)
Comment (fegin) [reply torch/distributed/_tensor/experimental/attention.py:375]: Yes, for the accuracy purpose. Otherwise, we may get NAN.

File: torch/distributed/_tensor/experimental/attention.py (hunk 4 above)
Comment (fegin) [reply torch/distributed/_tensor/experimental/attention.py]: I merged and remove the APIs. Now we only have one public API, `context_parallel()`.

File: torch/distributed/_tensor/experimental/attention.py (hunk 5 above)
Comment (fegin) [reply torch/distributed/_tensor/experimental/attention.py]: I merged and remove the APIs. Now we only have one public API, `context_parallel()`.

File: torch/distributed/distributed_c10d.py (hunk 3 above)
Comment (fegin) [reply torch/distributed/distributed_c10d.py:4938]: Oh, yes, the original one is not `torch.compiler()` friendly.

File: torch/distributed/_tensor/experimental/attention.py [hunk 7]
```diff
@@ -634,27 +724,136 @@ def backward_hook(grad: torch.Tensor) -> None:
...
+    ) -> Tuple[Tuple[Any, ...], Dict[str, Any]]:
+        placement = [Shard(seq_dim)]
+        all_args = []
 
-def _is_causal_behavior(
-    rank: int, world_size: int, i: int, is_causal: bool
-) -> _CausalBehavior:
+        for arg in itertools.chain(args, kwargs.values()):
+            if isinstance(arg, torch.Tensor) and not isinstance(arg, DTensor):
+                arg = DTensor.from_local(arg, mesh, placement, run_check=False)
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/attention.py:739]: not necessary in this PR, but I think we can add another check
```
if isinstance(arg, DTensor) and arg.placements[0] != placement:
    arg = arg.redistribute(placements=placement)
```

File: torch/distributed/_tensor/experimental/attention.py [hunk 8]
```diff
@@ -634,27 +724,136 @@ def backward_hook(grad: torch.Tensor) -> None:
...
+            positional embedding buffers. These buffers must be sharded along
+            the sequence dimension to ensure the accuracy. The sharding will
+            happen in-place, the buffer's shape will change within the context.
+            The buffers will be restored after the context finishes.
+            ``no_restore_buffers`` can be used to specify which buffers don't
+            need to be restored. Note that ``buffers`` should not contain any
+            nn.Parameter.
+        buffer_seq_dims (Optional[List[int]]): the sequence dimensions of ``buffers``.
+        no_restore_buffers (Optional[Set[torch.Tensor]]): buffers in these set
+            won't be restored after the context exists. This set must be a subset
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/attention.py:825]:
```suggestion
            won't be restored after the context exits. This set must be a subset
```

File: torch/distributed/_tensor/experimental/attention.py [hunk 9]
```diff
@@ -634,27 +724,136 @@ def backward_hook(grad: torch.Tensor) -> None:
...
+            of ``buffers``.
+
+    .. warning::
+        `torch.distributed._tensor.experimental.attention.context_parall` is a
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/attention.py:829]:
```suggestion
        `torch.distributed._tensor.experimental.attention.context_parallel` is a
```

File: torch/distributed/_tensor/experimental/attention.py [hunk 10]
```diff
@@ -634,27 +724,136 @@ def backward_hook(grad: torch.Tensor) -> None:
...
+            on the sequence dimension. Examples are input batch, labels and
+            positional embedding buffers. These buffers must be sharded along
+            the sequence dimension to ensure the accuracy. The sharding will
+            happen in-place, the buffer's shape will change within the context.
+            The buffers will be restored after the context finishes.
+            ``no_restore_buffers`` can be used to specify which buffers don't
+            need to be restored. Note that ``buffers`` should not contain any
+            nn.Parameter.
+        buffer_seq_dims (Optional[List[int]]): the sequence dimensions of ``buffers``.
+        no_restore_buffers (Optional[Set[torch.Tensor]]): buffers in these set
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/attention.py:824]: When user would determine whether the buffers should be restored or not? iiuc input batch/labels does not need to be restored? I wonder if there's a implication that this context manager need to be applied during the whole model fwd/bwd?

**Review by wanchaol — APPROVED** [review]
Nice work! I have a few more suggestions and questions inlined

File: torch/distributed/_tensor/ops/_matrix_ops.py [hunk 11]
```diff
@@ -360,6 +362,20 @@ def scaled_dot_product_efficient_attention_strategy(
     ]
     if has_attn_bias:
         all_replicate.append(Replicate())  # attn bias
+
+    # Context Parallelism: shards on the sequence dim
+    single_mesh_dim_strategies.append(
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/_matrix_ops.py:367]: I thought about this more, I think we should not include the context parallel strategy to the sharding strategy list here by default. The reason is that: context parallel strategy (shard on seq dim) is not a mathmatically correct strategy without ring attention being called locally. But if we add the strategy here, then the inputs with the strategy listed here would go through without ring attention being called, which could result in the wrong numerical results.

File: torch/distributed/_tensor/ops/_matrix_ops.py [hunk 12]
```diff
@@ -453,6 +469,27 @@ def scaled_dot_product_efficient_attention_backward_strategy(
     num_heads_dim_sharding.extend([Replicate(), Replicate()])
     single_mesh_dim_strategies.append(num_heads_dim_sharding)
 
+    # Context Parallelism: shards on the sequence dim
+    seq_dim_sharding: PlacementList = [
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/_matrix_ops.py:473]: ditto: we should not include the strategy here. I think what we should do instead:

since ring attention are currently op handlers, inside the op handlers, we don't need to run sharding propagation, as we know exactly the output sharding is, so we just create a `Shard(2)` output DTensorSpec directly before calling `DTensor._op_dispatcher.wrap(local_results, output_sharding.output_spec)`

File: torch/distributed/_tensor/experimental/attention.py (hunk 10 above)
Comment (fegin) [reply torch/distributed/_tensor/experimental/attention.py:824]: ye, this is mostly for efficiency, so batch and label don't need to be restored. And yes, the context manager must be applied during the whole fwd/bwd.

File: torch/distributed/_tensor/ops/_matrix_ops.py (hunk 11 above)
Comment (fegin) [reply torch/distributed/_tensor/ops/_matrix_ops.py:367]: good point. I agree we should not have the strategy by default. But I think we still need the propagation rule so that once we integrate with TP, the output sharding can be correctly inferred. I will have a separate PR to fix this.

File: torch/distributed/_tensor/experimental/attention.py [hunk 13]
```diff
@@ -634,27 +724,136 @@ def backward_hook(grad: torch.Tensor) -> None:
...
+            all_args.append(arg)
+
+        new_args = tuple(all_args[0 : len(args)])
+        new_kwargs = dict(zip(kwargs.keys(), all_args[len(args) :]))
+        return new_args, new_kwargs
+
+    def attention_output_fn(mesh: DeviceMesh, outputs: Any) -> Any:
+        new_outputs = []
+        for output in [outputs] if isinstance(outputs, torch.Tensor) else outputs:
+            output = output.to_local() if isinstance(output, DTensor) else output
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:750]: for my education, what is our assumption about user-original inputs/outputs before `context_parallel` wrapper gets involved?

it looks a little asymetric:
  - input_fn says "if args are not DTensor, make them DTensor sharded on seq_dim"
   - output_fn says "if output is a DTensor, unpack its local part
But how does output_fn know if user already provided a DTensor as input and expects a DTensor back at output vs a local part?

Comment (fegin) [reply torch/distributed/_tensor/experimental/attention.py:750]: Yes, so this part currently only support FSDP + CP and also with TorchTitan TP because in TorchTitan we convert it to local tensor before calling SDPA. But for completeness, this must be handled correctly in the `attention_output_fn`, so we can support the general case where we don't convert the DTensor (TP/SP) to the local tensor. That one I'm still working on.

**Comment by fegin** [comment]
Add a TODO to track the propagation rule issue.
